android/app/src/main/kotlin/net/mullvad/mullvadvpn/viewmodel/PaymentViewModel.kt

```diff
@@ -0,0 +1,42 @@
+package net.mullvad.mullvadvpn.viewmodel
+
+import android.app.Activity
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapNotNull
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+import net.mullvad.mullvadvpn.compose.dialog.payment.PaymentDialogData
+import net.mullvad.mullvadvpn.lib.payment.model.ProductId
+import net.mullvad.mullvadvpn.lib.payment.model.PurchaseResult
+import net.mullvad.mullvadvpn.usecase.PaymentUseCase
+import net.mullvad.mullvadvpn.util.toPaymentDialogData
+
+class PaymentViewModel(
+    private val paymentUseCase: PaymentUseCase,
+) : ViewModel() {
+    val uiState: StateFlow<PaymentUiState> =
+        paymentUseCase.purchaseResult
+            .mapNotNull { it?.toPaymentDialogData() }
+            .map { PaymentUiState(it) }
+            .stateIn(viewModelScope, SharingStarted.Lazily, PaymentUiState(PaymentDialogData()))
+
+    val uiSideEffect =
+        paymentUseCase.purchaseResult
+            .filter { it is PurchaseResult.Completed.Cancelled }
+            .map { PaymentUiSideEffect.PaymentCancelled }
```

Comment:
I guess the code is slightly confusing, and it might needs some test, but I think it might be that there a case where we get an error while in the Google payment flow that also should just end silently like cancel.

Reply:
It has been resolved!